### PR TITLE
chore: establish a basic style guide

### PR DESCRIPTION
**Pull request by tobz**
This PR attempts to establish the beginnings of a basic style guide. This style guide would dictate, primarily, what our preferred/allowed code style is, as well as establishing an opinionated stance on everything from formatting of log messages, to disallowed libraries, to common code patterns.

The guide is by no means done as of this PR: I mostly wanted to get the groundwork laid because part of reviewing this PR is also reviewing the yet-to-be-majority-approved _opinions_ that I've bootstrapped the guide with. In and of itself, I've added a lot of copy to be reviewed, so there's reviewing the actual wording and then reviewing the _decision_ I've inherently made on various topics... so adding even more stuff didn't seem prudent at this point. :P

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]

**Review by blt — APPROVED** [review]
I like it.

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+`std::sync` directly, which will be easier to switch to if we're already using `once_cell`.
+
+If you're working with data that changes over time, but has a very high read-to-write ratio, such as many readers, but
+one writer and infrequent writes, we prefer **[`arc-swap`](https://docs.rs/arc-swap)**.  The main feature of this crate
+is allowing a piece of data to be atomically updated while being shared concurrently. It does this by wrapping all data
+in `Arc<T>` to provide the safe, concurrent access, while adding the ability to atomically swap the `Arc<T>` itself. As
+it cannot be constructed in a const fashion, `arc-swap` pairs well with `once_cell` for actually storing it in a global
+static variable.
+
+### Concurrent data structures
```
Comment (bruceg) [inline STYLE.md:256]: If we're going to talk about these, would it also be worth mentioning `Mutex` vs `RwLock` and the std vs parking_lot vs tokio implementations?

**Review by bruceg — COMMENTED** [review]
Good start, I have some questions as usual.

File: STYLE.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+# Formatting
+
+At a high-level, code formatting is straightforward: we use the native `rustfmt` exclusively.
+
+You can acquire `rustfmt` -- which is invoked as `cargo fmt` -- by following the directions listed out [in the rustfmt
+repository](https://github.com/rust-lang/rustfmt#on-the-stable-toolchain).
+
+Vector has its own formatting rules (`.rustfmt.toml`) that will automatically be used when you run `cargo fmt` within
+the repository.  If you're ever in doubt, you can also run `make check-fmt` which will invoke `cargo fmt` and ensure all
+code follows our formatting rules.
```
Comment (StephenWakely) [inline STYLE.md:19]: `rustfmt` doesn't always format code within macros, in which case you should make a best effort to make the code formatted like the code in the rest of the project.

File: STYLE.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+We use `lib` almost entirely for shared libraries and for isolating specific pieces of code. As Vector itself involves a
+large number of dependencies, it can be beneficial to move code into isolated crates under `lib/` in order to allow them
+not only to be shared, but also to reduce the amount of code that must be processed by helper tools like `cargo check`
+and `rust-analyzer` during normal development, which in turn speeds up the feedback loop between writing code and
+getting informed about errors, warnings, and so on.
+
+## `src/`: main binary and all related functionality
+
+The bulk of code resides in the `src/` directory/crate, and within that, is almost entirely related to the
```
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:35]: Yeah, I guess the thrust of the statement is meant to refer to the user-facing code -- components, configuration parsing, and so on. I'll think about how to better reword this to be along the lines of a code organization _map_: rather than how much code lives where, more of a... "if you're looking for X, you should look here", etc.

File: STYLE.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+As a large, open-source project, it can be a struggle to ensure a consistent level of code quality.  This style guide is
+meant to be the canonical reference for all things style: code comments, formatting, acceptable (or unacceptable)
+crates, data structures, or algorithms, and so on.
+
+In essence, we hope to turn pull request review comments like "why did you do it this way?" or "I think you could try
+doing it this way" into "we always do X this way: <link to style guide>".
+
+# Formatting
+
+At a high-level, code formatting is straightforward: we use the native `rustfmt` exclusively.
```
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:12]: What do you mean by "... and on all sources."?

File: STYLE.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+```rust
+// Don't do this:
+info!(message = "Something happened.");
+// Do this instead:
+info!("Something happened.");
+
+// Don't do this:
+debug!(%client_id, message = "Client entered authentication phase.");
+debug!(message = "Client entered authentication phase.", %client_id);
```
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:85]: Aha, so, first debate about one of my baseline opinions that I shoved in here. :)

My _personal_ feeling is that using `message = "Foo bar doot doot doot..."` is just extra characters to type and clutters up the expression. It looks especially weird when there's no structured data attached, just a "normal" log, as it doesn't match the macros of `log`, even though they can be used interchangeably for non-structured logging.

That said, I didn't realize -- per your `check-events` comment -- that we're looking for instances of `message = "..."` in order to assert correct grammar on them. That does throw a wrench into this particular decree. I'll marinate on this one and see if _maybe_ there's a simple tweak to `check-events` to easily handle my proposed version.

Still interested in thoughts about, ignoring the ability to parse the form in `check-events`, the proposed form and if there's pros/cons, or just opposing personal opinions, that I haven't considered.

File: STYLE.md (hunk 1 above)
Comment (tobz) [reply STYLE.md:256]: It probably would be a good idea, yes. I'll slot something in.

File: STYLE.md (hunk 2 above)
Comment (tobz) [reply STYLE.md:19]: That's a good point. I think we could maybe add a note to be on the lookout, although I don't want people to think they always _have to be_ on the lookout. Ideally it'd be something we could lightly suggest in reviews if we spot it.

Either way, I'll try to add a small note or reword it to indicate that `cargo fmt` is the bare _minimum_ but there might need to be other manual tweaks done to fit within the expected formatting style.

Comment (StephenWakely) [reply STYLE.md:19]: Yeah, it's just something that surprised me a couple of times when I started doing Rust. I don't we need any more than a small note to point out that it's because the code is in a macro that the formatting is not working as expected.

File: STYLE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply STYLE.md:12]: I meant to suggest saying that we use `rustfmt` both exclusively and comprehensively, that is on all Rust source code files.

File: STYLE.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply STYLE.md:85]: `check-events` looks for both plain strings and `message = "…"` strings, so that's not the issue per se. The problem is, it only searches at the very front of the log construct for that magic. I _think_ I can adapt the regex to look for either location, but it'll get hairy. It's also worth noting that the actual output puts the message before the structured fields (following the implicit stuff `log` provides, of course).

[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:85]: Coming back to this: I haven't yet taken a look at tweaking the regex in `check-events`, but one thing I did look at is using positional format arguments in the message string.

When using the `message = "..."` format, you _cannot_ use positional arguments like you could when using `println!` or `format!`. This lends a little bit of an edge, IMO, to always using the `"message without the field name", ...` approach, because it always works regardless of there are structured fields before it, or positional arguments after it... whereas doing `message = "...", positional_arg` will simply treat `positional_arg` as a structured field, and emit the format string as-is.

File: STYLE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md]: Addressed in ae129d597.

File: STYLE.md (hunk 2 above)
Comment (tobz) [reply STYLE.md]: Addressed in ae129d597.

File: STYLE.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:97]: I've taken a pass at reworking the `check-events` script, specifically the log message validation logic, to accommodate for not only the codebase as it stands now, but also to handle the format I'm proposing in the style guide.

File: STYLE.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+// Do this instead:
+debug!(%client_id, "Client entered authentication phase.");
```
[the remark this answers is not in the document]
Comment (tobz) [reply STYLE.md:99]: I've made changes to `check-events` to handle this, and other, cases.

File: STYLE.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+Vector has its own formatting rules (`.rustfmt.toml`) that will automatically be used when you run
+`cargo fmt` within the repository.  If you're ever in doubt, you can also run `make check-fmt` which
+will invoke `cargo fmt` in a dry-run mode, checking to see if any changed files are not currently
+formatted correctly.
+
+As an additional note, `rustfmt` sometimes can fail to format code within macros, so if you happen
+to see such code that doesn't look like it's formatted correctly, you may need to manually tweak it
+if `rustfmt` cannot be persuaded to format it correctly for you. :)
+
```
Comment (StephenWakely) [inline STYLE.md:26]: Could we add a section on code documentation as per here: https://github.com/vectordotdev/vector/blob/master/docs/REVIEWING.md?plain=1#L76-L89

**Review by spencergilbert — APPROVED** [review]
It's probably worth adding some notes about writing good documentation as @StephenWakely noted - I'm happy to leave that as a todo item though.

File: STYLE.md (hunk 7 above)
Comment (tobz) [reply STYLE.md:26]: Created #14573 to track doing so as a follow-up item.
